packages/compat/src/transaction.ts

```diff
@@ -0,0 +1,158 @@
+import { Base58EncodedAddress } from '@solana/addresses';
+import { pipe } from '@solana/functional';
+import { AccountRole, IAccountMeta, IInstruction } from '@solana/instructions';
+import { Ed25519Signature } from '@solana/keys';
+import {
+    appendTransactionInstruction,
+    Blockhash,
+    createTransaction,
+    ITransactionWithBlockhashLifetime,
+    ITransactionWithFeePayer,
+    ITransactionWithSignatures,
+    setTransactionFeePayer,
+    setTransactionLifetimeUsingBlockhash,
+    Transaction,
+} from '@solana/transactions';
+import {
+    MessageAccountKeys,
+    MessageCompiledInstruction,
+    PublicKey,
+    VersionedMessage,
+    VersionedTransaction,
+} from '@solana/web3.js';
+
+function convertAccount(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    accountIndex: number
+): IAccountMeta {
+    const accountPublicKey = accountKeys.get(accountIndex);
+    if (!accountPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find account address at index ${accountIndex}`);
+    }
+    const isSigner = message.isAccountSigner(accountIndex);
+    const isWritable = message.isAccountWritable(accountIndex);
+
+    const role = isSigner
+        ? isWritable
+            ? AccountRole.WRITABLE_SIGNER
+            : AccountRole.READONLY_SIGNER
+        : isWritable
+        ? AccountRole.WRITABLE
+        : AccountRole.READONLY;
+
+    return {
+        address: accountPublicKey.toBase58() as Base58EncodedAddress,
+        role,
+    };
+}
+
+// While the type `MessageCompiledInstruction` says data is a `Uint8Array`,
+// it is actually a Buffer in tests. Presumably this is a bug in the old web3js compile
+// message code. We will just assume it can be either here
+type MessageCompiledInstructionButDataMightBeABuffer = Omit<MessageCompiledInstruction, 'data'> & {
+    data?: Uint8Array | Buffer;
+};
```

Comment:
Buffer is a sub-type of Uint8Array!

Reply:
Hmm, that'd explain this annoyance!
The old one takes a `Buffer`, then compiles it to `Uint8Array` but actually just returns the buffer (which I guess isn't a bug given this sub-type relationship)
But in the tests the `Buffer` ends up as a different and I think worse representation:

```diff
-     "data": Uint8Array [
+     "data": Object {
+       "data": Array [
          0,
          1,
          2,
          3,
          4,
        ],
+       "type": "Buffer",
+     },
```

Maybe it doesn't really matter since it's a `Uint8Array` anyway and we should just leave it as a Buffer?

Reply:
Alright I've just dropped this, it's typed as a `Uint8Array` so we shouldn't care that under the hood it's a `Buffer`. No point adding complexity just to make the tests the exact shape I expected, consumers shouldn't care.

Reply:
The reason that people lean on `Buffer` is because it has this convenience that `Uint8Array` does _not_:

```ts
buf.toString('base64'); // Actually works.
```

So long as you can presume that peculiar feature is not needed on the way in or out of here, you can type it `Uint8Array` and it'll be fine.